### PR TITLE
ref(models): `update_or_create()`

### PR DESCRIPTION
**Pull request by mgaeta**
Introduce a replacement for the deprecated `create_or_update()`. Separating this from https://github.com/getsentry/sentry/pull/32335 so that https://github.com/getsentry/getsentry/pull/7134 can access the function.

### CODE REVIEW

**Comment by ceorourke** [comment]
I don't think our team owns this code, since every model may use this it makes me nervous. Maybe @getsentry/owners-api can review? I'm not sure if that's the right team but it seems more correct than ecosystem.

File: src/sentry/db/models/query.py [hunk 1]
```diff
@@ -49,6 +66,50 @@ def update(self: Model, using: str | None = None, **kwargs: Any) -> int:
 update.alters_data = True  # type: ignore
 
 
+def update_or_create(
```
Comment (wedamija) [inline src/sentry/db/models/query.py:69]: One difference between this and the old function is that we used to be able to differentiate between defaults and values for updating. So defaults would only be used on create, where values would be used on both create/update.

I'm not sure if we have any code that relies on that behaviour. If not, then I'm find with dropping it, but let's double check.

**Review by wedamija — COMMENTED** [review]
Changes to update look good to me. This generally seems like a good change, just a couple comments.

Also worth noting that we're already using `update_or_create` in a few places. Probably fine since this should also just be a drop in.

File: src/sentry/db/models/query.py [hunk 2]
```diff
@@ -11,30 +11,47 @@
...
 
 
-def update(self: Model, using: str | None = None, **kwargs: Any) -> int:
+def _handle_value(instance: Model, value: Any) -> Any:
+    if isinstance(value, CombinedExpression):
+        return resolve_combined_expression(instance, value)
+    return value
+
+
+def _handle_key(model: Type[Model], key: str, value: Any) -> str:
```
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/db/models/query.py:27]: In this PR `_handle_key()` is only used in `update_or_create()`. I think in the previous PR I was also using it in `create_or_update()` before I delete it.

File: src/sentry/db/models/query.py [hunk 3]
```diff
@@ -49,6 +66,50 @@ def update(self: Model, using: str | None = None, **kwargs: Any) -> int:
...
+
+    create_kwargs.update(
+        {_handle_key(model, k, v): _handle_value(instance, v) for k, v in defaults.items()}
+    )
+
+    try:
+        with transaction.atomic(using=using):
+            return objects.create(**create_kwargs), True
+    except IntegrityError:
+        pass
```
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/db/models/query.py:108]: I think the race condition of a third update coming between this functions's two `update()` calls is actually worth avoiding. If this function's updates would happen at  `t = 0` and `t = 2`, then the `t = 1` update from another call would be overwritten even though it was received last.

File: src/sentry/db/models/query.py (hunk 1 above)
Comment (mgaeta) [reply src/sentry/db/models/query.py:69]: OK I double checked and the only calls that were using both `values` and `defaults` were:
```py
GroupEnvironment.objects.create_or_update(
    environment_id=caches["Environment"](project.organization_id, env_name).id,
    group_id=group_id,
    defaults=fields,
    values=fields,
)
```

```py
rows_affected, created = OrganizationOnboardingTask.objects.create_or_update(
    organization=organization, task=task_id, values=values, defaults={"user": request.user}
)
```
```py
OrganizationMemberTeam.objects.create_or_update(
    organizationmember=to_member, team=omt.team, defaults={"is_active": True}
)
```
The first one is equivalent to omitting `defaults`, the second one just for the product tutorial, and the third one should probably be replaced with a `create()`.

File: src/sentry/db/models/query.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/db/models/query.py:108]: Are there any cases where we've passed `F('field') + 1`?  That's one concern I have here here.

I also don't think this solves the race condition you're talking about. This change just chooses first write wins vs last write wins. Neither is particularly more correct, but one is behaviour that the system has had for many years.

File: src/sentry/db/models/query.py (hunk 1 above)
Comment (wedamija) [reply src/sentry/db/models/query.py:69]: For the second this does change behaviour, but maybe it's not too important...

File: src/sentry/db/models/query.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mgaeta) [reply src/sentry/db/models/query.py:108]: Cool I think I get it now 👍🏽  
We only do that in mocks, tests, and getsentry's usage receiver `record_metric_usage()`:
```py
BillingMetricUsage.objects.create_or_update(
    billing_history=history,
    billing_metric=filters["billing_metric"],
    date=filters["date"],
    outcome=filters["outcome"],
    project_id=filters["project_id"],
    reason_code=filters["reason_code"],
    values={c: F(c) + v for c, v in columns.items()},
)
```
I'll add back the update with a little note acknowledging why we're retrying.

**Review by wedamija — APPROVED** [review]

**Comment by wedamija** [comment]
As a follow up pr let's put this in the manager to override the existing `update_or_create` that Django provides, then migrate existing uses of `create_or_update` in a 3rd

**Pull request merged**